alembic/testing/env.py

```diff
@@ -90,10 +87,53 @@ def env_file_fixture(txt):
         f.write(txt)
 
 
-def _sqlite_file_db(tempname="foo.db"):
-    dir_ = os.path.join(_get_staging_directory(), "scripts")
-    url = "sqlite:///%s/%s" % (dir_, tempname)
-    return engines.testing_engine(url=url)
+if win:
+    # windows cannot remove an open file. An sqlite db file stays open
+    # until any connection to it is open, so we need to keep track of the
+    # engine and of all the connection to close them when clearing the
+    # the staging env
+    def clear_staging_env():
+        for entry in _engine_registry.values():
+            _clear_engine(entry)
+        _engine_registry.clear()
+        shutil.rmtree(_get_staging_directory(), True)
+
+    _engine_registry = {}
+
+    def _clear_engine(entry):
+        entry["engine"].dispose()
+        for conn in entry["connections"]:
+            conn.close()
+
+    def _sqlite_file_db(tempname="foo.db"):
+        dir_ = os.path.join(_get_staging_directory(), "scripts")
+        url = "sqlite:///%s/%s" % (dir_, tempname)
+        eng = engines.testing_engine(url=url)
+
+        if url in _engine_registry:
+            _clear_engine(_engine_registry[url])
+        entry = {"engine": eng, "connections": []}
+        _engine_registry[url] = entry
+        _orig_connect = eng.connect
+
+        def connect(*args, **kwargs):
+            conn = _orig_connect(*args, **kwargs)
+            entry["connections"].append(conn)
+            return conn
+
+        eng.connect = connect
+        return eng
```

Comment:
I'm not sure if there is a more elegant way to get all the connections. This version seems to work but it less than elegant

Reply:
oh.   right, so yeah, there's no need to monkeypatch, there's an event system that intercepts these.    within the testing suite, this should already be happening, that is, if you call:

    from sqlalchemy.testing import engines
    engines.testing_reaper.close_all()

this should be getting called after every test suite teardown.     it's getting called right here:

https://github.com/sqlalchemy/alembic/blob/master/alembic/testing/plugin/plugin_base.py#L28

so basically none of the above should be needed.

Reply:
by which I mean, if we are getting a stack trace there and it is failing to remove the files, some test or fixture is not correctly using the testing.engines package to track that engine.

Reply:
The stack trace is on "foo.db". Cannot remove open file.

The stack trace happens on basically every test that uses the db when using sqlite

Reply:
One of the classes that has the error is this one `tests/test_command.py::StampMultipleHeadsTest`

It calls only `clear_staging_env` in its tear down

> this should be getting called after every test suite teardown. it's getting called right here: 

this seems to only work once at the end of the class, not after every test.
So it seems that `clear_staging_env` runs after every test, but the engine clear runs only at the end of the class.

I'll use `engines.testing_reaper.close_all()` in `clear_staging_env`